File: leetcode/top_interview_150/Hashmap/383. Ransom Note/main.go
package main

import "fmt"

func canConstruct(ransomNote string, magazine string) bool {
	magazMap := make(map[rune]int)
	for _, let := range magazine {
		magazMap[let] += 1
	}

	for _, let := range ransomNote {
		if count, ok := magazMap[let]; ok && count > 0 {
			magazMap[let] -= 1
			continue
		} else {
			return false
		}
	}

	return true
}

func main() {
	fmt.Println(canConstruct("a", "b"))
	fmt.Println(canConstruct("aa", "ab"))
	fmt.Println(canConstruct("aa", "aab"))
}

File: leetcode/top_interview_150/Hashmap/383. Ransom Note/main.py
class Solution(object):
    def canConstruct(self, ransomNote: str, magazine: str) -> str:
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """

        mapped = {}
        for el in magazine:
            if mapped.get(el):
                mapped[el] += 1
            else:
                mapped[el] = 1

        for let in ransomNote:
            if mapped.get(let, False) and mapped[let] > 0:
                mapped[let] -= 1
                continue
            else:
                return False
        return True


if __name__ == "__main__":
    print(Solution().canConstruct("a", "b"))
    print(Solution().canConstruct("aa", "ab"))
    print(Solution().canConstruct("aa", "aab"))
